app/src/main/java/com/example/someapplication/data/MoviesRepository.kt

```diff
@@ -11,51 +17,101 @@ import kotlinx.coroutines.withContext
 class MoviesRepository {
 
     private val api = NetworkModule().provideApiService()
+    private val daoMovieDetails = AppDataBase.instance.getMovieDetailsDbDao()
+    private val daoMoviesList = AppDataBase.instance.getMoviesListDbDao()
 
-    suspend fun loadMovies(): List<MoviePreview>? {
-        var moviePreviewList: List<MoviePreview>? = null
+    suspend fun loadMovies(): List<MoviesListEntity>? {
+        var moviePreviewListDto: List<MoviePreview>? = null
+        var moviesResult: List<MoviesListEntity>? = null
         try {
-            moviePreviewList = withContext(Dispatchers.IO) {
+            moviePreviewListDto = withContext(Dispatchers.IO) {
                 api.loadMovies(KEY, LANG, PAGE).results
             }
+            moviesResult = Mapper.mapMoviesListToDb(moviePreviewListDto)
+            withContext(Dispatchers.IO) {
+                daoMoviesList.deleteCachedMoviesList()
+                daoMoviesList.saveMoviesList(moviesResult)
+            }
         } catch (e: Exception) {
         }
-        return moviePreviewList
+        return moviesResult
     }
 
-    suspend fun getGenres(): List<Genre>? {
-        var genres: List<Genre>? = null
+    suspend fun getGenres(): List<GenresEntity>? {
+        var genresDto: List<Genre>? = null
+        var genresResult: List<GenresEntity>? = null
         try {
-            genres = withContext(Dispatchers.IO) {
+            genresDto = withContext(Dispatchers.IO) {
                 api.getGenres(KEY, LANG).genres
             }
+            genresResult = Mapper.mapGenresToDb(genresDto)
+            withContext(Dispatchers.IO) {
+                daoMoviesList.deleteCachedGenresList()
+                daoMoviesList.saveGenresList(genresResult)
+            }
         } catch (e: Exception) {
         }
-        return genres
+        return genresResult
+    }
+
+    suspend fun loadCachedMovies(): List<MoviesListEntity>? {
+        return withContext(Dispatchers.IO) {
+            daoMoviesList.getCashedMoviesList()
+        }
+    }
+
+    suspend fun getCachedGenres(): List<GenresEntity>? {
+        return withContext(Dispatchers.IO) {
+            daoMoviesList.getCashedGenresList()
+        }
     }
 
-    suspend fun getMovie(movieId: Int): MovieFull? {
+    suspend fun getMovie(movieId: Int): MovieDetailsEntity? {
         var movieFull: MovieFull? = null
+        var movieDetailsEntity: MovieDetailsEntity? = null
         try {
             movieFull = withContext(Dispatchers.IO) {
                 api.getMovie(movieId, KEY, LANG)
             }
+            movieDetailsEntity = Mapper.mapDetailsToDb(movieFull)
+            withContext(Dispatchers.IO) {
```

Comment:
I think, You don't need to use withContext every time, because under the hood Room like Retrofit uses a more effective dispatcher for work.